File: packages/template-php/src/Source.php
<?php

declare(strict_types=1);

namespace Polyspec\Template;

/**
 * Template source: UTF-8 validation, BOM removal and the line index (LEX-1, LEX-2, LEX-16).
 */
final class Source
{
    /** @var list<int> byte offsets of line starts */
    public readonly array $lines;

    private function __construct(public readonly string $name, public readonly string $text)
    {
        $this->lines = self::lineIndex($text);
    }

    public static function fromBytes(string $name, string $bytes): self
    {
        $invalid = Utf8::firstInvalid($bytes);
        if ($invalid >= 0) {
            throw TemplateError::at('E_LEX_INVALID_UTF8', $name, self::lineIndex($bytes), $invalid, $invalid + 1, "invalid UTF-8 byte at offset {$invalid}");
        }
        if (str_starts_with($bytes, "\xEF\xBB\xBF")) {
            $bytes = substr($bytes, 3);
        }

        return new self($name, $bytes);
    }

    /**
     * @return list<int>
     */
    public static function lineIndex(string $text): array
    {
        $starts = [0];
        $offset = 0;
        while (($found = strpos($text, "\n", $offset)) !== false) {
            $starts[] = $found + 1;
            $offset = $found + 1;
        }

        return $starts;
    }

    /**
     * @param list<int> $lines
     * @return array{0: int, 1: int} 1-based line and byte column
     */
    public static function position(array $lines, int $offset): array
    {
        $low = 0;
        $high = count($lines) - 1;
        while ($low < $high) {
            $mid = intdiv($low + $high + 1, 2);
            if ($lines[$mid] <= $offset) {
                $low = $mid;
            } else {
                $high = $mid - 1;
            }
        }

        return [$low + 1, $offset - $lines[$low] + 1];
    }

    public function length(): int
    {
        return strlen($this->text);
    }
}

File: packages/template-rust/src/source.rs
//! Template source: UTF-8 validation and BOM removal (LEX-1, LEX-2). Offsets are byte offsets.

use crate::error::{ErrorCode, LineIndex, TemplateError};

/// A validated template source.
#[derive(Debug, Clone)]
pub struct Source {
    /// Template name used in errors and in the AST.
    pub name: String,
    /// Source text after BOM removal.
    pub text: String,
    /// Line index of the text.
    pub lines: LineIndex,
}

impl Source {
    /// Builds a source from bytes; rejects invalid UTF-8 and removes a leading BOM.
    pub fn from_bytes(name: &str, input: &[u8]) -> Result<Source, TemplateError> {
        if let Err(error) = std::str::from_utf8(input) {
            let invalid = error.valid_up_to();
            return Err(TemplateError::at(
                ErrorCode::E_LEX_INVALID_UTF8,
                name,
                Some(&LineIndex::of(input)),
                [invalid, invalid + 1],
                format!("invalid UTF-8 byte at offset {invalid}"),
            ));
        }
        let bytes = if input.starts_with(&[0xef, 0xbb, 0xbf]) {
            &input[3..]
        } else {
            input
        };
        let text = String::from_utf8(bytes.to_vec()).expect("validated");
        Ok(Source {
            name: name.to_string(),
            lines: LineIndex::of(bytes),
            text,
        })
    }

    /// Builds a source from a string; removes a leading BOM.
    pub fn from_text(name: &str, text: &str) -> Source {
        let stripped = text.strip_prefix('\u{feff}').unwrap_or(text);
        Source {
            name: name.to_string(),
            lines: LineIndex::of(stripped.as_bytes()),
            text: stripped.to_string(),
        }
    }

    /// The bytes of the text.
    pub fn bytes(&self) -> &[u8] {
        self.text.as_bytes()
    }

    /// Creates an error at a byte span of this source.
    pub fn error(&self, code: ErrorCode, start: usize, end: usize, message: impl Into<String>) -> TemplateError {
        TemplateError::at(code, &self.name, Some(&self.lines), [start, end], message)
    }
}
